[JENKINS-40958] Allow Pull Request Approved Builds

kylec32 opened the pull request:
This pull request allows `pullrequest:approved` webhook requests to trigger a build in the plugin. 
Also extended some of the objects coming back from the Bitbucket API to include more data about reviewers and approvals to allow extension plugins to operate with additional information. 

By itself this PR does basically nothing. If a user of the plugin enabled "Approved" pull request webhook requests in Bitbucket it would potentially trigger additional builds in this plugin; however, that would require a manual action by the user. 

The primary use case for this plugin is to enable the building of an extension plugin that would allow Pull request filtering based on if a PR had  been approved or not thus fulfilling the objective of Jenkins 40958. This functionality is being implemented in an extension plugin and not as part of this plugin as was suggested in PR 44 to this repository. However I would hope the additions could also assist in other extension plugins/future features of this plugin.

michaelneale commented:
very interesting looking - I think I get what this is about, don't feel qualified to code review bit it seems pretty neat to me.

kylec32 commented:
Any feedback on this? Happy to make any requested changes. 
Thanks.

amuniz commented:
Is there a public plugin using this?

kylec32 commented:
There will be. Haven't really made it public since it's useless without this.

michaelneale commented:
@kylec32 is there some repo that can be referred to that shows it (even if not publicised)?

kylec32 commented:
@michaelneale @amuniz No problem, here is how I intend to use it: https://github.com/kylec32/bitbucket-scm-filter-approval/tree/master

Thank you for taking a look.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketReviewer.java:
```diff
@@ -0,0 +1,40 @@
+     * @return the username of the reviewer.
+     */
+    String getReviewerLogin();
+}
```
Missing newline at the end of the file

kylec32 replied:
Should have the newline now.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketPullRequestFull.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Represents all the details of a pull request.
+ */
+public interface BitbucketPullRequestFull extends BitbucketPullRequest {
```
Is this really necessary? Why not just add the `getReviewers` to the `BitbucketPullRequest` class

amuniz replied:
Yeah, I think it could be added to the existing mapping class.

kylec32 replied:
I totally see how this could look weird so let me explain. 

Right now the plugin pulls in all the pull requests for a particular repo with a url such as:
`https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests{?page,pagelen}`
This returns all the data contained in the `BitbucketPullRequest` interface for every PR in that repo. However, it does not return any data about reviewers/approvers in that response. In order to get that you need to call a different endpoint along the lines of: 
`https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests/{id}` (notice that we are accessing a single PR ID)
This is the endpoint that the `getPullRequestById()` function calls and thus includes the reviewer information. (This function is never called within this plugin as far as I'm aware currently.)
Taking this all into account I created the new interface so that everything that currently exists would have the same interface to interact with and not have to worry that `getReviewers()` would always return null and `getPullRequestById()` could return the full version since it does have access to all the data. 

Hopefully the reasoning is a bit clearer now. I definitely wish I didn't have to do this but alternatives were along the lines of making an additional HTTP call for each pull request to get the additional data and that seemed much worse.

amuniz replied:
Ok, so we need a new HTTP call. Still I would keep the client API object _hidden_, my proposal:

1. Add `BitbucketSCMSourceRequest#getPullRequestById(String id)`
1. It will use [`source.buildBitbucketClient()`](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/fc0785417c4cdc6afc777f88c2447cfeb8266f2c/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java#L508) to create a client and make the HTTP call to get the PR details - note that the `BitbucketSCMSource` is [passed to the request](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/2379c9b5a581d09688567e45de66b4f86f822de8/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java#L126).
1. Use `BitbucketSCMSourceRequest#getPullRequestById(String id)` in the extension plugin.
1. The implementation of `BitbucketSCMSourceRequest#getPullRequestById(String id)` should use some sort of caching mechanism. Two consecutive calls to `getPullRequestById` for the same PR should do no more than one HTTP request. The lifecycle of the `BitbucketSCMSourceRequest` object is bounded to a specific `SCMSource#retrieve` call so the cache won't need any _invalidate_, _erase_ nor _eviction_ handling (a plain old `HashMap<String, BitbucketPullRequestFull>` should be enough.

This way `BitbucketApi` remains _private_ and extensions can get PR details by using the already available `BitbucketSCMSourceRequest` object.

WDTY?

kylec32 replied:
Seems reasonable to me. I'll implement that when I get a chance and update here. Thank you for the suggestion.

rsandell replied:
Until we get java 8 adding more interfaces just for a new method is dangerous, what to do next time we want to add a new method? New interface? what should it be called? `BitbucketPullRequestFuller`? :)
Since we don't have default methods yet, maybe a new interface is needed, but it could simply be called `BitbucketPullRequest2`. wdyt?

kylec32 replied:
So the reasoning behind the name is, as somewhat explained above, there are two sources of pull request metadata. The first source, and the one primarily used in this plugin, is the `https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests{?page,pagelen}` endpoint. This of course gets data about multiple pull requests at a time which leads to a reduced number of HTTP requests required to get the data. However this endpoint only returns a subset of the data about a pull request, one of the pieces it doesn't include is reviewer information. That comes only from the other source of pull request metadata which is `https://api.bitbucket.org/2.0/repositories/{/owner,repo}/pullrequests/{id}` this only returns data about one pull request but includes all the data about the pull request. This is why I chose the name, because this is the place where you get the full information about a pull request. So if we had a new method we wanted to add and the source of it's data was not one of the above two endpoints we could indeed be in an annoying situation with naming. My bet with the naming of this was that it wouldn't happen though. 
I'm fine to change the name if there is passion behind it but I personally like `BitbucketPullRequestFull` the most of any option I have thought of.

casz replied:
Still don't think is needed: You could settle for `@JsonInclude(Include.NON_EMPTY)` on `getReviewers`

amuniz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java:
```diff
@@ -115,7 +116,10 @@
+    /**
+     * The BitbucketApi that is used for the request.
+     */
+    private BitbucketApi api;
```
I think this is not needed.

AFAICT this is intended to be [used by the extension to recover the PR info](https://github.com/kylec32/bitbucket-scm-filter-approval/blob/master/src/main/java/tk/lightweightcoding/jenkins/bitbucket/approval/ApprovedPullRequestTrait.java#L100). However required info can be just added to the original `BitbucketPullRequest` (as suggested by @casz) and then used in the extension. This way no new API call is needed.

Also, I'd like to keep `BitbucketApi` as _private_ as possible. I know it is technically public but as currently designed, extensions do not have the need to use it (and it should remain like that).

kylec32 replied:
I do see the desire to keep this as private as possible but as described in the comment above the `BitbucketPullRequest` that is passed to me doesn't have the information I need (reviewers) and I don't have access to the credentials or anything like that to make the request myself so this is what I ended up with.

Definitely open for suggestions.

tarioch commented:
Sorry if I'm abusing this PR but as you're already adding more info to the PR would it make sense to also add the canMerge info to the PR? On the one hand querying for that information would fix JENKINS-45997 and on the other hand it could be handy to also limit builds for stuff which has canMerge true.

That way it would then be possible to only build if it can be merged and is approved, therefore reducing the amount of "useless" builds.

casz commented:
@tarioch that would still require a trait extension plugin once you have the data.

tarioch commented:
Created PR #116 for the first part (calling the can merge api) and once this is done will look at creating a trait which uses this information.

kylec32 commented:
The requested changes should now be applied.

kylec32 commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/pullrequest/BitbucketPullRequestValue.java:
```diff
@@ -37,6 +39,10 @@
 
     private Author author;
 
+    //Bitbucket cloud seems to have all the same information in the participant field plus more.
```
Yes, most of the testing has in fact been on Bitbucket Cloud.

kylec32 commented:
@amuniz I think I'm basically all set to release my trait. 
https://github.com/jenkinsci/bitbucket-approval-filter-plugin
https://wiki.jenkins.io/display/JENKINS/Bitbucket+Approval+Filter+Plugin
https://github.com/jenkins-infra/repository-permissions-updater/pull/647 <- Little worried this is going to be rejected as my trait will break if someone were to try to use it without this plugins changes (hence it's recent failing builds)
This plugin will have to release first since mine has a dependency on it and once I have a release I can depend on I should be able to push my plugin.

casz commented:
kylec just don't make a release until the plugin is released

kylec32 commented:
Understood. I'm ready to release as soon as this is merged in and released so just let me know if there is anything else holding up this pull request. Thank you.

rsandell commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketReviewer.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Represents the reviewer of a pull request.
+ */
+public interface BitbucketReviewer {
```
Is this really needed? Something in the json parser that requires it? Seems to me like a standard pojo would be enough otherwise.

kylec32 replied:
There are two implementations of this interface. `BitbucketReviewerValue` for Bitbucket Cloud and `BitbucketServerReviewer` for Bitbucket Server. The way that `getReviewerLogin` is implemented is different in cloud and server unfortunately because the field names are different in the JSON (`user.name` vs `user.username`).

casz replied:
Not needed, we can update to Jackson 2.9 and use [`JsonAlias`](https://www.concretepage.com/jackson-api/jackson-jsonproperty-and-jsonalias-example#JsonAlias)

rsandell commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/pullrequest/BitbucketPullRequestValue.java:
```diff
@@ -93,6 +99,15 @@ public String getAuthorDisplayName() {
     }
 
+    @Override
+    public List<BitbucketReviewerValue> getReviewers() {
```
You don't seem to need the `BitbucketReviewer` interface here?

kylec32 replied:
`BitbucketPullRequestFull`'s contract is that it will return a list of something that extends from `BitbucketReviewer` (`List<? extends BitbucketReviewer>`) and thus here we are returning a list of `BitbucketReviewerValue` and in `BitbucketServerPullRequest` we are returning a list of `BitbucketServerReviewer`. So while we aren't using `BitbucketReviewer` explicitly here we are implementing an interface that uses it. 
Hopefully I didn't misinterpret your question there.

kylec32 commented:
Is there anything else that is holding up this merge? Particularly @rsandeell were there any other questions or concerns with this change that I could address? Thank you.

casz commented:
I went ahead and updated jackson in #143 so you can use `JsonAlias` 😇

kylec32 commented:
@casz I have rebased so there are no longer any conflicts. I liked the suggestion for using `@JsonAlias` although it also appears that the update to Jackson 2.9 had to be reverted 😢 I changed how that code was working in order to handle the differences between Bitbucket cloud and server differently. It's not as elegant as I would have liked but it works. 

Any other feedback?

casz commented:
@kylec32 you could try and use the incremental release of Jackson 2.9.7 as a showcase for using Java 8 in this library 🙌

in the POM.xml you switch `jackson2-api` to `2.9.7.0-rc70.f9e409daf2bb`
```xml
   <dependency>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>jackson2-api</artifactId>
     <version>2.9.7.0-rc70.f9e409daf2bb</version>
   </dependency>
```
and switch `java.level` to `8` unde properties.

Then you should be able to develop with JSON Alias and I don't we will see a release with this before we switch to Java 8 anyway!

kylec32 commented:
@casz Updated code to now use `@JsonAlias` (to the beta version released today). Also increased the java version to 8.

You mentioned that you don't see this PR getting merged until this plugin switches to Java 8. Is there a timeline when this is planned to happen?

casz commented:
@kylec32 as soon as git and git-client goes into java 8 🙌 which should happen soonish as they are in beta 😅

casz commented:
hi @kylec32 you have some checkstyle erros :)

casz commented:
My bad I caused them :cry:

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java:
```diff
@@ -113,6 +114,19 @@
+    /**
+     * A map serving as a cache of pull request IDs to the full set of data about the pull request.
+     */
+    private final Map<Integer, BitbucketPullRequest> pullRequestData;
```
I'd like something that cleans the cache

casz replied:
Fair enough, my bad :frowning:

casz commented:
Thanks for added the native webhook processor support! :sweat_smile:

casz commented:
@kylec32 
you might want to look into fixing your git user.email config 😅 
or set up an email alias on your GitHub account 🥇 
to get proper credit for your commits 🏆 
https://help.github.com/articles/adding-an-email-address-to-your-github-account/
https://help.github.com/articles/setting-your-commit-email-address-in-git
then to amend your commits: https://stackoverflow.com/a/3042512

kylec32 commented:
Good catch with the email alias. That's now fixed 😄.

As far as I know I have made all the changes needed. As you likely saw I changed it slightly to use `account_id` instead of `username` as the user identifier when talking to Bitbucket cloud as username is going to be gone after the April GDPR updates 👋. I also updated my filter plugin to use the latest changes from this PR. 

Anything else need to be done before a merge 🙏?

casz commented:
:tada:

kylec32 commented:
Yay! 🕺 🎉 👍 

Thank you!

casz commented:
Released :smile: Hopefully, you can get your plugin hosted. https://wiki.jenkins.io/display/JENKINS/Hosting+Plugins
